File: trees/tree_traversal.js
/* Tree traversal algorithms using recursion in JavaScript */

/* Node class */
class Node {
    // Constructor
    constructor(data) {
        this.data = data
        this.left = null
        this.right = null
    }
    /* Methods */

    // Inorder traversal algorithm
    inorderTraversal(root) {
        // Check if there's no root, simply return
        if (!root) {
            return
        }
        // Otherwise, implement the inorder traversal algorithm using recursion
        this.inorderTraversal(root.left)
        console.log(`The tree is: ${root.data}`)
        this.inorderTraversal(root.right)
    }

    // Preorder traversal algorithm
    preorderTraversal(root) {
        // Check if there's no root, simply return
        if (!root) {
            return
        }
        // Otherwise, implement the preorder traversal algorithm using recursion
        console.log(`The tree is: ${root.data}`)
        this.preorderTraversal(root.left)
        this.preorderTraversal(root.right)
    }
    // Postorder traversal algorithm
    postorderTraversal(root) {
        // Check if there's no root, simply return
        if (!root) {
            return
        }
        // Otherwise, implement the postorder traversal algorithm using recursion
        this.postorderTraversal(root.left)
        this.postorderTraversal(root.right)
        console.log(`The tree is: ${root.data}`)

    }
}

// Instantiate the node class by creating a few instances
root = new Node(5)
root.left = new Node(7)
root.left.right = new Node(20)
root.left.left = new Node(8)
root.left.left.right = new Node(15)
root.left.left.left = new Node(10)
root.right = new Node(30)
root.right.left = new Node(50)
root.right.right = new Node(65)
root.right.right.left = new Node(100)
root.right.right.right = new Node(90)
root.inorderTraversal(root)
root.preorderTraversal(root)
root.postorderTraversal(root)

File: trees/tree_traversal.py
# Tree traversal algorithm in Python using recursino

# Node class
class Node:
    # Constructor
    def __init__(self, data):
        self.data = data
        self.left = None
        self.right = None

    # Methods

    # Inorder traversal
    def inorder_traversal(self, root):
        # Check if the root is None
        if root == None:
            return
        # Otherwise implement the inorder traversal algorithm recursively
        self.inorder_traversal(root.left)
        print(f'The tree is: {root.data}', end=' ')
        self.inorder_traversal(root.right)

    # Preorder traversal
    def preorder_traversal(self, root):
        # Check if the root is None
        if root == None:
            return
        # Otherwise implement the preorder traversal algorithm recursively
        print(f'The tree is: {root.data}', end=' ')
        self.preorder_traversal(root.left)
        self.preorder_traversal(root.right)

    # Postorder traversal
    def postorder_traversal(self, root):
        # Check if the root is None
        if root == None:
            return
        # Otherwise implement the postorder algorithm recursively
        self.postorder_traversal(root.left)
        self.postorder_traversal(root.right)
        print(f'The tree is: {root.data}', end=' ')


if __name__ == '__main__':
    # Instantiate the node class by creating objects (nodes)
    node = Node(5)
    node.left = Node(7)
    node.left.right = Node(20)
    node.left.left = Node(8)
    node.left.left.right = Node(15)
    node.left.left.left = Node(10)
    node.left.left.right = Node(15)
    node.right = Node(30)
    node.right.left = Node(50)
    node.right.right = Node(65)
    node.right.right.left = Node(100)
    node.right.right.right = Node(90)
    node.inorder_traversal(node)
    print()
    node.preorder_traversal(node)
    print()
    node.postorder_traversal(node)
    print()
